circuit/src/gadgets/set.rs

```diff
@@ -279,6 +279,142 @@ impl<F: FieldExt> Chip<F> for SetPositionChip<F> {
+			//  |------------|---------|--------|-----------|-----------|--------|-------|-------|-------|
+			//  |            |         |        |           |           |        |       |       |   0   |
+			//  |     *      |    2    |    0   |     2     |   2^(-1)  |   0    |   1   |   0   |   0   |
+			//  |     *      |    2    |    1   |     1     |   1^(-1)  |   0    |   2   |   0   |   0   |
+			//  |     *      |    2    |    2   |     0     |   1       |   1    |   3   |   3   |   3   |
+			//  |     *      |    2    |    3   |    -1     |  -1^(-1)  |   0    |   4   |   0   |   3   |
+			//  |     *      |    2    |    4   |    -2     |  -2^(-1)  |   0    |   5   |   0   |   3   |
+			//
+
+			let idx_exp = v_cells.query_advice(common.advice[0], Rotation::cur());
+			let id_exp = v_cells.query_advice(common.advice[1], Rotation::cur());
+			let diff_exp = v_cells.query_advice(common.advice[2], Rotation::cur());
+			let diff_inv_exp = v_cells.query_advice(common.advice[3], Rotation::cur());
+			let select_exp = v_cells.query_advice(common.advice[4], Rotation::cur());
+			let elem_exp = v_cells.query_advice(common.advice[5], Rotation::cur());
+			let add_exp = v_cells.query_advice(common.advice[6], Rotation::cur());
+
+			let item_prev_exp = v_cells.query_advice(common.advice[7], Rotation::prev());
+			let item_exp = v_cells.query_advice(common.advice[7], Rotation::cur());
+
+			let s_exp = v_cells.query_selector(selector);
+
+			let one_exp = Expression::Constant(F::ONE);
+
+			vec![
+				// If the difference is equal to 0, that will make the `select` equal to 1.
+
+				// diff = idx(target) - id(elem)
+				s_exp.clone() * (idx_exp - (diff_exp.clone() + id_exp)),
+				// select * (1 - select) = 0
+				s_exp.clone() * (select_exp.clone() * (one_exp.clone() - select_exp.clone())),
+				//
+				// if diff = 0 { select = 1 } else { select = 0 }
+				// 		diff * (1 / diff) = 1 - select
+				//      1 - select - diff * diff_inv = 0
+				//
+				s_exp.clone() * (one_exp - select_exp.clone() - diff_exp * diff_inv_exp),
+				// add = select * elem
+				s_exp.clone() * (add_exp.clone() - (select_exp * elem_exp)),
+				// item = add + item_prev
+				s_exp * (item_exp - (item_prev_exp + add_exp)),
+			]
+		});
+
+		selector
+	}
+
+	/// Synthesize the circuit.
+	fn synthesize(
+		self, common: &CommonConfig, selector: &Selector, mut layouter: impl Layouter<F>,
+	) -> Result<Self::Output, Error> {
+		layouter.assign_region(
+			|| "select_item_with_idx",
+			|region: Region<'_, F>| {
+				let mut ctx = RegionCtx::new(region, 0);
+
+				let mut item_prev_cell = ctx.assign_from_constant(common.advice[7], F::ZERO)?;
+				ctx.next();
+
+				let mut assigned_item = ctx.assign_from_constant(common.advice[7], F::ZERO)?;
+				let mut assigned_target_idx =
+					ctx.copy_assign(common.advice[0], self.idx.clone())?;
+				for i in 0..self.items.len() {
+					ctx.enable(*selector)?;
+
+					let id_cell = ctx.assign_from_constant(common.advice[1], F::from(i as u64))?;
```

Comment:
I think we should use fixed column for this.
Underneath, this uses both advice and a fixed column, so we can make it a little bit cheaper by just using fixed column

Reply:
Hmm, I tried the update you recommended.
There is a small issue.
When doing the unit tests, the `production` test fails with `synthesis` error.
Normal `MockProver` test is okay.
I am digging into it.

Reply:
I think we should use the `advice` column for `id` because it is variant which depends on `items` - circuit input.
Specifically, `id` is the column for assigning the index of `items`. 
So, I think it is variant, and `advice` column should be used.

Reply:
It fails because a production prover goes through the circuit 2 times. First at compile time when fixed columns are assigned, then again at proving time where fixed columns are not assigned and values returned are Value::unassigned().
Since you are using this value here: https://github.com/eigen-trust/protocol/blob/select-item-chip/circuit/src/gadgets/set.rs#L381, it gives you error.

Reply:
You can do this to solve it:
```rust
let id = F::from(i as u64);
ctx.assign_fixed(common.fixed[1], id)?;

// Calculating difference between given target idx and id of element from the set.
let diff = self.idx.value().cloned() - Value::known(id);
```

Reply:
Or you could keep it as advice, but constrain it to be incremental (1 higher then previous row), and not use fixed column

Reply:
Ok, but keep this in mind when we get to optimising the circuit layout at the end, in case these fixed columns stop us from having better layout